# A12.1 (a)

In [6]:
using LinearAlgebra
A = randn(20,10)

20×10 Matrix{Float64}:
 -2.22594    -0.730973   -0.532295  …  -1.61737    -1.84818    1.21025
 -0.554591   -0.468117    2.21878       0.530708   -0.435208  -0.371985
 -0.0955603  -0.114154   -0.857354      0.18077     0.512513  -0.0415953
  0.291862    0.394806   -0.395661      0.315001   -0.394096  -0.717001
  0.150994    0.72153     0.303278      0.766009   -1.39881    0.127032
 -0.471512   -0.795096    1.73565   …  -0.898204    0.151703  -0.689566
 -1.28093     1.36352    -2.36618      -0.0771968   2.42042    0.0637638
  0.912047    0.495617   -0.586587      0.998792    0.953727   0.984028
  1.1472     -2.94529     1.667         1.11369    -0.861296   0.496303
  0.550802    1.34032     0.508817      2.71619    -0.546906   0.029466
  0.214882   -0.445897   -0.628175  …  -1.21474     0.970407   1.69917
  0.625659   -0.407881    0.310379      1.25075     0.65807   -0.711622
 -0.435258    1.19208     2.37195      -0.429083    1.1581     1.09157
 -0.53549     0.838528   -1.00839      -0.

In [3]:
b = randn(20,1)

20×1 Matrix{Float64}:
  0.9131859659001478
 -0.6669801741011908
 -1.206061352246283
  0.4730096991845767
 -0.2425439583540643
  0.40176924184215856
 -1.9233693397566594
  0.1711942934973721
 -0.5472670574914832
  0.3007063010074322
  2.283763536357731
  1.5673271520065593
 -1.113614928128954
  1.2177144932761566
 -1.6475754950451842
  0.15502656688850908
  1.4277890005619465
  0.5836655285181357
  0.4849647053452569
 -0.6744481628434149

In [4]:
x̂_1 = A\b
val_x̂_1=norm(A*x̂_1-b)^2

14.021788446320775

In [5]:
x̂_2 = inv(A'*A)*(A'*b)
norm(A*x̂_2-b)^2

14.021788446320775

In [6]:
x̂_3=pinv(A)*b
norm(A*x̂_3-b)^2

14.021788446320775

### As we can see all values of x̂ generated using different methods are very close to each other.

# A12.1 (b)

In [7]:
δ = randn(10,1)

10×1 Matrix{Float64}:
  1.582865367736138
 -0.7297125448612269
 -0.06977852509169732
 -0.47509348698456827
 -0.43963046253682103
  0.8227488947365944
  0.3589410538877154
 -1.378440835349625
 -1.0771831101509002
 -0.8895051850958401

In [8]:
δ_1 = δ*10^-2
δ_2 = δ*10^-3
δ_3 = δ*10^-4
δ_4 = δ*10^-5
δ_5 = δ*10^-6

10×1 Matrix{Float64}:
  1.5828653677361385e-6
 -7.297125448612273e-7
 -6.977852509169735e-8
 -4.7509348698456843e-7
 -4.396304625368212e-7
  8.227488947365947e-7
  3.5894105388771553e-7
 -1.3784408353496254e-6
 -1.0771831101509006e-6
 -8.895051850958404e-7

In [9]:
val_1=norm(A*(x̂_1 + δ_1) - b)^2
val_2=norm(A*(x̂_1 + δ_2) - b)^2
val_3=norm(A*(x̂_1 + δ_3) - b)^2
val_4=norm(A*(x̂_1 + δ_4) - b)^2
val_5=norm(A*(x̂_1 + δ_5) - b)^2

14.021788446441786

#### we have considered val_x̂_1 from the previous question A12.1(a)

In [10]:
val_1 > val_x̂_1

true

In [11]:
val_2 > val_x̂_1

true

In [12]:
val_3 > val_x̂_1

true

In [13]:
val_4 > val_x̂_1

true

In [14]:
val_5 > val_x̂_1

true

### Thus we can see that the expected inequatity  ||A(x̂ + δ) − b||^2  > ||Ax̂ − b||^2 is verified.

# A12.2

In [15]:
A = randn(100000,100)

100000×100 Matrix{Float64}:
 -1.21475    -0.563822   -0.429748   …  -0.071057   -1.1685     -0.84826
  0.506539    0.396134   -0.796634       0.0169777   1.35287    -1.03468
  0.217337    0.686375   -1.06568       -0.653217    0.447492    1.80255
  0.812946   -0.232145   -1.05285       -0.0605392   0.870079    0.221127
  1.29096     1.15951     0.188939      -1.4307     -0.988712    1.48136
 -0.292039    0.435821    0.21608    …  -0.501165    2.80849     1.42283
 -1.96125    -0.665909    0.0126434      0.899118   -1.6424     -0.403573
  1.55109    -0.933064    0.227414       0.653643   -0.608385   -0.195192
  0.422884    0.0676013   0.528501       0.224412    0.799214   -0.170018
  0.484221    0.0356477  -0.131675       2.74185    -0.949901    2.19405
  0.833182    0.673738    1.79588    …  -0.365133    0.464319    0.922109
  0.464494    0.683182   -0.0417324     -1.52525     0.0102295  -0.147723
  1.64135     1.74095     1.31802        0.380576    1.03256     0.531603
  ⋮             

In [16]:
b = randn(100000)

100000-element Vector{Float64}:
 -1.0301668734380798
  0.9679328799591805
 -1.1633884738586562
 -0.05878883490358528
 -0.10653056399469296
 -0.10344044358421359
  0.4068539098413889
  1.565736857013875
  0.17710272491429965
 -0.38973410912898876
 -0.22897515672761637
  0.13538320509452803
  2.075590768065514
  ⋮
  0.8045324762254549
 -0.9359735943765058
 -0.8348875445971343
 -0.530297321177133
  2.2685030116334306
 -0.4901892029610973
  0.582131654537154
  0.32742326658207654
 -0.7325961785451893
 -1.1357517496011023
  0.8648162956150454
  1.2719379481930093

In [17]:
@time begin
A\b
end

  1.002065 seconds (442.43 k allocations: 102.827 MiB, 8.87% gc time, 24.18% compilation time)


100-element Vector{Float64}:
 -0.0017773812670047642
  0.0025984462910533584
  0.0004805016201897422
  0.002584068104499633
 -0.005362284412017402
 -0.0019479938622248025
  0.002567557673313238
 -0.0002644961626308565
  0.002612643147388151
  0.0033108296932892996
  0.000545737173701712
  0.0022330385192751738
  0.003884399506901462
  ⋮
  0.0007033522655967322
  0.0010245183258775967
 -0.003183379630257167
 -0.0033091891044435095
 -0.004846991685315471
  0.0009908054536667047
  4.367303446254721e-5
  0.0006372602523754101
  0.0044203125639191
 -0.003565161599786098
 -0.0024495221341000388
  0.00010475782233330922

In [18]:
@time begin
A\b
end

  0.712282 seconds (633 allocations: 77.364 MiB, 7.95% gc time)


100-element Vector{Float64}:
 -0.0017773812670047642
  0.0025984462910533584
  0.0004805016201897422
  0.002584068104499633
 -0.005362284412017402
 -0.0019479938622248025
  0.002567557673313238
 -0.0002644961626308565
  0.002612643147388151
  0.0033108296932892996
  0.000545737173701712
  0.0022330385192751738
  0.003884399506901462
  ⋮
  0.0007033522655967322
  0.0010245183258775967
 -0.003183379630257167
 -0.0033091891044435095
 -0.004846991685315471
  0.0009908054536667047
  4.367303446254721e-5
  0.0006372602523754101
  0.0044203125639191
 -0.003565161599786098
 -0.0024495221341000388
  0.00010475782233330922

In [19]:
@time begin
A\b
end

  0.728635 seconds (633 allocations: 77.364 MiB)


100-element Vector{Float64}:
 -0.0017773812670047642
  0.0025984462910533584
  0.0004805016201897422
  0.002584068104499633
 -0.005362284412017402
 -0.0019479938622248025
  0.002567557673313238
 -0.0002644961626308565
  0.002612643147388151
  0.0033108296932892996
  0.000545737173701712
  0.0022330385192751738
  0.003884399506901462
  ⋮
  0.0007033522655967322
  0.0010245183258775967
 -0.003183379630257167
 -0.0033091891044435095
 -0.004846991685315471
  0.0009908054536667047
  4.367303446254721e-5
  0.0006372602523754101
  0.0044203125639191
 -0.003565161599786098
 -0.0024495221341000388
  0.00010475782233330922

In [20]:
@time begin
A\b
end

  0.791196 seconds (633 allocations: 77.364 MiB)


100-element Vector{Float64}:
 -0.0017773812670047642
  0.0025984462910533584
  0.0004805016201897422
  0.002584068104499633
 -0.005362284412017402
 -0.0019479938622248025
  0.002567557673313238
 -0.0002644961626308565
  0.002612643147388151
  0.0033108296932892996
  0.000545737173701712
  0.0022330385192751738
  0.003884399506901462
  ⋮
  0.0007033522655967322
  0.0010245183258775967
 -0.003183379630257167
 -0.0033091891044435095
 -0.004846991685315471
  0.0009908054536667047
  4.367303446254721e-5
  0.0006372602523754101
  0.0044203125639191
 -0.003565161599786098
 -0.0024495221341000388
  0.00010475782233330922

In [21]:
@time begin
A\b
end

  0.704723 seconds (633 allocations: 77.364 MiB, 0.32% gc time)


100-element Vector{Float64}:
 -0.0017773812670047642
  0.0025984462910533584
  0.0004805016201897422
  0.002584068104499633
 -0.005362284412017402
 -0.0019479938622248025
  0.002567557673313238
 -0.0002644961626308565
  0.002612643147388151
  0.0033108296932892996
  0.000545737173701712
  0.0022330385192751738
  0.003884399506901462
  ⋮
  0.0007033522655967322
  0.0010245183258775967
 -0.003183379630257167
 -0.0033091891044435095
 -0.004846991685315471
  0.0009908054536667047
  4.367303446254721e-5
  0.0006372602523754101
  0.0044203125639191
 -0.003565161599786098
 -0.0024495221341000388
  0.00010475782233330922

### Julia took 0.54-0.56 seconds to solve a least squares problem with m = 100000 equations and n = 100 variables on my computer.

# A13.3 (b)

In [7]:
include("time_series_data.jl")

function toeplitz(vect,n) ==>
	forms the toeplitz matrix as described in section 7.4 in textbook
	n is the size of the vector to convolve with
	select the rows you need from this

data given is x_train and x_test


100-element Vector{Float64}:
  0.140126610206199
 -0.3266617955319319
 -0.3190213361122029
 -0.07285822610013504
  0.10625789758760604
  0.06124284356474724
 -0.46306782099841665
  0.13028759972404974
  0.3285691916684773
 -0.04611914392286799
 -0.29126645858125566
 -0.1336195642051234
 -0.11155462013022835
  ⋮
  1.329897476359537
  1.5076467764325698
 -1.2476869129326498
 -1.4682439883434155
  0.9317559762963956
  1.5366874780961386
 -1.1309996959748272
 -1.588768005489042
  0.6139908129155712
  1.8193852647716864
 -0.4834015865194686
 -1.9298247132201785

In [8]:
N=100 #size of x_train and x_test is known
for M in 2:12
    xtrain=toeplitz(x_train,M)
    xtest=toeplitz(x_test,M)
    xtrain=xtrain[M:N-1,:]
    xtest=xtest[M:N-1,:]
    b=x_train[M+1:N]
    b_test=x_test[M+1:N]
    β=xtrain\b
    print("\n_______________________________M=",M,"_______________________________","\n")
    print("Model parameters for M=",M," are : ",round.(β,sigdigits=4),"\n")
    print("Mean square prediction error for train data:",round.(norm(xtrain*β - b)^2/(N-M),sigdigits=4))
    print("\nMean square prediction error for test data:",round.(norm(xtest*β - b_test)^2/(N-M),sigdigits=4))
    print("\nDifference between mean square error valiues for train and test data:",(norm(xtest*β - b_test)^2/(N-M)-norm(xtrain*β - b)^2/(N-M)))
end


_______________________________M=2_______________________________
Model parameters for M=2 are : [0.1082, -0.9802]
Mean square prediction error for train data:0.03049
Mean square prediction error for test data:0.03874
Difference between mean square error valiues for train and test data:0.00824963132226808
_______________________________M=3_______________________________
Model parameters for M=3 are : [0.05642, -0.9722, -0.05097]
Mean square prediction error for train data:0.03029
Mean square prediction error for test data:0.03743
Difference between mean square error valiues for train and test data:0.007140842850674369
_______________________________M=4_______________________________
Model parameters for M=4 are : [0.08788, -0.5379, -0.07048, 0.4382]
Mean square prediction error for train data:0.0245
Mean square prediction error for test data:0.03386
Difference between mean square error valiues for train and test data:0.009359393160680195
_______________________________M=5_____________

|M|Mean square prediction error(J)|
|---|---|
|M=2|Train error = 0.03049, Test error = 0.03874|
|M=3|Train error = 0.03029, Test error = 0.03743|
|M=4|Train error = 0.0245, Test error = 0.03386|
|M=5|Train error = 0.02439, Test error = 0.03169|
|M=6|Train error = 0.02299, Test error = 0.03040|
|M=7|Train error = 0.01984, Test error = 0.03524|
|M=8|Train error = 0.01914, Test error = 0.0346|
|M=9|Train error = 0.01908, Test error = 0.03705|
|M=10|Train error = 0.01899, Test error = 0.03768|
|M=11|Train error = 0.01920, Test error = 0.03815|
|M=12|Train error = 0.01892, Test error = 0.04030|

### As per the error values in the above table the most suitable value for M is 6 since the model has the least mean square prediction error on x_test (i.e. the test data that is used to validate the model), at M=6.

In [46]:
M=1
xtrain=toeplitz(x_train,M)
xtrain=xtrain[M:N-1,:]
b=x_train[M+1:N]
β=1
print("J value for simplew predictor (1):",round.(norm(xtrain*β - b)^2/(N-M),sigdigits=4))

J value for simplew predictor (1):2.651

In [47]:
M=2
xtrain=toeplitz(x_train,M)
xtrain=xtrain[M:N-1,:]
b=x_train[M+1:N]
β=[2,-1]
print("J value for simplew predictor (2):",round.(norm(xtrain*β - b)^2/(N-M),sigdigits=4))

J value for simplew predictor (2):5.021